File: src/mrtjp/projectred/core/Configurator.java
package mrtjp.projectred.core;

import cpw.mods.fml.client.event.ConfigChangedEvent;
import cpw.mods.fml.common.event.FMLPreInitializationEvent;
import cpw.mods.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.common.config.Configuration;

import java.io.File;

public class Configurator
{
    protected static File _configFolder;
    protected static File _configFile;
    public static Configuration config;

    /** Constants **/
    public static final String modName = "Project Red";
    public static final String version = "@VERSION@";
    public static final String buildnumber = "@BUILD_NUMBER@";

    /** Generation **/
    public static boolean gen_MarbleCave;
    public static int gen_MarbleCave_resistance;
    public static boolean gen_MarbleCave_retro;
    public static boolean gen_Volcano;
    public static int gen_Volcano_resistance;
    public static boolean gen_Volcano_retro;
    public static boolean gen_Ruby;
    public static int gen_Ruby_resistance;
    public static boolean gen_Ruby_retro;
    public static boolean gen_Sapphire;
    public static int gen_Sapphire_resistance;
    public static boolean gen_Sapphire_retro;
    public static boolean gen_Peridot;
    public static int gen_Peridot_resistance;
    public static boolean gen_Peridot_retro;
    public static boolean gen_SpreadingMoss;
    public static boolean gen_Lily;
    public static int gen_Lily_resistance;
    public static boolean gen_Lily_retro;
    public static boolean gen_Copper;
    public static int gen_Copper_resistance;
    public static boolean gen_Copper_retro;
    public static boolean gen_Tin;
    public static int gen_Tin_resistance;
    public static boolean gen_Tin_retro;
    public static boolean gen_Silver;
    public static int gen_Silver_resistance;
    public static boolean gen_Silver_retro;
    public static boolean gen_Electrotine;
    public static int gen_Electrotine_resistance;
    public static boolean gen_Electrotine_retro;

    /** Settings **/
    public static boolean versionChecking;
    public static boolean versionCheckDevBuilds;

    public static boolean logicGateSounds;
    public static int minTimerTicks;

    public static boolean simpleFramedWireRecipe;

    public static int maxDetectionCount;
    public static int detectionFrequency;
    public static int routerUpdateThreadCount;

    public static int maxPipesWandered;

    /** Render **/
    public static boolean logicwires3D;
    public static boolean staticWires;
    public static boolean staticGates;
    public static int lightHaloMax;
    public static boolean coloredLightsCompat;

    public static void initConfig(FMLPreInitializationEvent event)
    {
        _configFolder = event.getModConfigurationDirectory();
        _configFile = new File(_configFolder.getAbsolutePath() + "/ProjectRed.cfg");
        config = new Configuration(_configFile);
        updateConfig();
    }

    public static void updateConfig()
    {
        config.addCustomCategoryComment("World Generation", "Toggle generation of structures, or increase resistance to lessen chances for generation.");
        gen_Ruby = config.get("World Generation", "Ruby Ore", true).getBoolean(true);
        gen_Ruby_resistance = config.get("World Generation", "Ruby Ore resistance", 0).getInt();
        gen_Ruby_retro = config.get("World Generation", "Ruby Ore retrogen", false).getBoolean(false);

        gen_Sapphire = config.get("World Generation", "Sapphire Ore", true).getBoolean(true);
        gen_Sapphire_resistance = config.get("World Generation", "Sapphire Ore resistance", 0).getInt();
        gen_Sapphire_retro = config.get("World Generation", "Sapphire Ore retrogen", false).getBoolean(false);

        gen_Peridot = config.get("World Generation", "Peridot Ore", true).getBoolean(true);
        gen_Peridot_resistance = config.get("World Generation", "Peridot Ore resistance", 0).getInt();
        gen_Peridot_retro = config.get("World Generation", "Peridot Ore retrogen", false).getBoolean(false);

        gen_MarbleCave = config.get("World Generation", "Marble Caves", true).getBoolean(true);
        gen_MarbleCave_resistance = config.get("World Generation", "Marble Caves resistance", 0).getInt();
        gen_MarbleCave_retro = config.get("World Generation", "Marble Caves retrogen", false).getBoolean(false);

        gen_Volcano = config.get("World Generation", "Volcanos", true).getBoolean(true);
        gen_Volcano_resistance = config.get("World Generation", "Volcano resistance", 0).getInt();
        gen_Volcano_retro = config.get("World Generation", "Volcano retrogen", false).getBoolean(false);

        gen_Lily = config.get("World Generation", "Deviating Lily", true).getBoolean(true);
        gen_Lily_resistance = config.get("World Generation", "Deviating Lily resistance", 0).getInt();
        gen_Lily_retro = config.get("World Generation", "Deviating Lily retrogen", false).getBoolean(false);

        gen_Copper = config.get("World Generation", "Copper Ore", true).getBoolean(true);
        gen_Copper_resistance = config.get("World Generation", "Copper Ore resistance", 0).getInt();
        gen_Copper_retro = config.get("World Generation", "Copper Ore retrogen", false).getBoolean(false);

        gen_Tin = config.get("World Generation", "Tin Ore", true).getBoolean(true);
        gen_Tin_resistance = config.get("World Generation", "Tin Ore resistance", 0).getInt();
        gen_Tin_retro = config.get("World Generation", "Tin Ore retrogen", false).getBoolean(false);

        gen_Silver = config.get("World Generation", "Silver Ore", true).getBoolean(true);
        gen_Silver_resistance = config.get("World Generation", "Silver Ore resistance", 0).getInt();
        gen_Silver_retro = config.get("World Generation", "Silver Ore retrogen", false).getBoolean(false);

        gen_Electrotine = config.get("World Generation", "Electrotine Ore", true).getBoolean(true);
        gen_Electrotine_resistance = config.get("World Generation", "Electrotine Ore resistance", 0).getInt();
        gen_Electrotine_retro = config.get("World Generation", "Electrotine Ore retrogen", false).getBoolean(false);

        gen_SpreadingMoss = config.get("World Generation", "Spreading Moss", true).getBoolean(true);

        versionChecking = config.get("general", "Enable Version checking", true, "If enabled, the player will be alerted if a newer version of PR is available").getBoolean(true);
        versionCheckDevBuilds = config.get("general", "Dev Version checking", false, "Include dev builds in version checks").getBoolean(false);

        logicGateSounds = config.get("general", "Logic Sounds", true, "If set to false, logic gates will not make sounds.").getBoolean(true);
        minTimerTicks = config.get("general", "Mininum Timer Ticks", 4, "Minimum amount of ticks the timer gates can be set to. (Cannot be lower than 4)").getInt();
        logicwires3D = config.get("general", "3Dlogicwires", true, "If set to false, flat wire textures will be used for logic gates. Significant performance improvement").getBoolean(true);
        staticWires = config.get("general", "renderStaticWires", true, "If set to false, wires will be rendered in the TESR rather than the WorldRenderer").getBoolean(true);
        staticGates = config.get("general", "renderStaticGates", true, "If set to false, gates will be rendered in the TESR rather than the WorldRenderer").getBoolean(true);
        lightHaloMax = config.get("general", "Light Halo Render Count", -1, "Number of lights to render, -1 for unlimited").getInt();
        coloredLightsCompat = config.get("general", "coloredLightsCompat", false, "Enable compatability with ColoredLightsCore. CLC is in beta state and is known to cause many glitches.").getBoolean(false);

        maxDetectionCount = config.get("general", "Max Detection Count", 100, "Max number of links to explore when discovering new routers.").getInt();
        detectionFrequency = config.get("general", "Detection Frequency", 20, "Ticks between router searches.").getInt();
        routerUpdateThreadCount = config.get("general", "Router Update Thread Count", 4, "Number of active threads that update routing tables.").getInt();

        maxPipesWandered = config.get("general", "Max Pipes Wandered", 0, "Maximum number of pipes an object can wander through before being erased. 0 for unlimited.").getInt();

        simpleFramedWireRecipe = config.get("general", "Simple Framed Wire recipe", false, "Use sticks instead of wood strips in framed wire recipes.").getBoolean();

        if (config.hasChanged())
            config.save();
    }

    @SubscribeEvent
    public void onConfigChanged(ConfigChangedEvent.OnConfigChangedEvent event)
    {
        if ("ProjRed|Core".equals(event.modID))
            updateConfig();
    }
}

File: src/core/scala/mrtjp/projectred/core/Configurator.scala
/*
 * Copyright (c) 2016.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.projectred.core

import mrtjp.core.data.ModConfig

object Configurator extends ModConfig("projectred-core") {
    /** Constants * */
    var modName = "Project Red"

    /** General Settings * */
    var logicGateSounds = true
    var minTimerTicks = 4
    var simpleFramedWireRecipe = false
    var unbreakableScrewdriver = false

    var maxDetectionCount = 100
    var detectionFrequency = 40
    var routerUpdateThreadCount = 4
    var maxPipesWandered = 0

    /** Machines * */
    var enableDiamondBlockBreaker = false

    /** Rendering * */
    var logicwires3D = true
    var staticWires = true
    var staticGates = true
    var lightHaloMax = -1
    var pipeRoutingFX = true

    /** World Gen * */
    var gen_MarbleCave = true
    //    var gen_MarbleCave_resistance = 0
    //    var gen_MarbleCave_retro = false
    //    var gen_Volcano = true
    //    var gen_Volcano_resistance = 0
    //    var gen_Volcano_retro = false
    var gen_Ruby = true
    //    var gen_Ruby_resistance = 0
    //    var gen_Ruby_retro = false
    var gen_Sapphire = true
    //    var gen_Sapphire_resistance = 0
    //    var gen_Sapphire_retro = false
    var gen_Peridot = true
    //    var gen_Peridot_resistance = 0
    //    var gen_Peridot_retro = false
    var gen_Copper = true
    //    var gen_Copper_resistance = 0
    //    var gen_Copper_retro = false
    var gen_Tin = true
    //    var gen_Tin_resistance = 0
    //    var gen_Tin_retro = false
    var gen_Silver = true
    //    var gen_Silver_resistance = 0
    //    var gen_Silver_retro = false
    var gen_Electrotine = true
    //    var gen_Electrotine_resistance = 0
    //    var gen_Electrotine_retro = false

    /** Compatibility * */
    //    var compat_Treecapitator = true
    //    var compat_CCBundledCalbe = true
    //    var compat_Chisel = true
    //    var compat_TConstruct = true
    //    var compat_TExpansion = true
    //    var compat_ColoredLights = false
    //    var compat_MFRDeepStorage = true

    override def getFileName = "ProjectRed"

    override protected def initValues() {
        val general = new BaseCategory("General Settings", "Contains basic settings for the mod.")

        logicGateSounds = general.put("Logic Sounds", logicGateSounds, "If set to false, logic gates will not make sounds.")
        minTimerTicks = general.put("Minimum Timer Ticks", minTimerTicks, "Minimum amount of ticks the timer gates can be set to. Cannot be lower than 4.")
        simpleFramedWireRecipe = general.put("Simple Framed Wire Recipe", simpleFramedWireRecipe, "If set to true, sticks will be used instead of wood strips in framed wire recipes.")
        unbreakableScrewdriver = general.put("Unbreakable Screwdriver", unbreakableScrewdriver, "If set to true, the basic screwdriver will not take damage.")


        maxDetectionCount = general.put("Routed Pipes: Max Detection Count", maxDetectionCount, "Max number of pipes to explore when connecting to other routers.")
        detectionFrequency = general.put("Routed Pipes: Detection Frequency", detectionFrequency, "Ticks between router searches.")
        routerUpdateThreadCount = general.put("Routed Pipes: Update Threads", routerUpdateThreadCount, "Number of active route table update threads.")
        maxPipesWandered = general.put("Routed Pipes: Max Wander Distance", maxPipesWandered, "Maximum distance an item may aimlessly wander in a pipe before being erased. 0 for unlimited.")

        val machines = new BaseCategory("Machine Settings", "Contains settings related to machines and devices.")
        enableDiamondBlockBreaker = machines.put("Enable the Diamond Block Breaker", enableDiamondBlockBreaker, "Allow the Diamond Block Breaker to be crafted.")

        val rendering = new BaseCategory("Render Settings", "Contains settings related to how things render in-game.")
        logicwires3D = rendering.put("3D Logic Wires", logicwires3D, "If set to false, flat wire textures will be used for logic gates. Significant performance improvement.")
        staticWires = rendering.put("Static Wires", staticWires, "If set to false, wires will be rendered in the TESR rather than the WorldRenderer.")
        staticGates = rendering.put("Static Gates", staticGates, "If set to false, gates will be rendered in the TESR rather than the WorldRenderer.")
        lightHaloMax = rendering.put("Light Halo Render Count", lightHaloMax, "Number of lights to render, -1 for unlimited")
        pipeRoutingFX = rendering.put("Routed Pipe FX", pipeRoutingFX, "If set to false, routed pipes will not render routing fx such as bubbles and lasers.")

        val gen = new BaseCategory("World Gen", "Contains settings related to world gen. You can enable/disable each ore or strucure, change retro generation settings, and increase how rare something is by increasing the resistance value.")
        gen_Ruby = gen.put("Ruby Ore", gen_Ruby)
        //        gen_Ruby_resistance = gen.put("Ruby Ore resistance", gen_Ruby_resistance)
        //        gen_Ruby_retro = gen.put("Ruby Ore retrogen", gen_Ruby_retro)
        //
        gen_Sapphire = gen.put("Sapphire Ore", gen_Sapphire)
        //        gen_Sapphire_resistance = gen.put("Sapphire Ore resistance", gen_Sapphire_resistance)
        //        gen_Sapphire_retro = gen.put("Sapphire Ore retrogen", gen_Sapphire_retro)
        //
        gen_Peridot = gen.put("Peridot Ore", gen_Peridot)
        //        gen_Peridot_resistance = gen.put("Peridot Ore resistance", gen_Peridot_resistance)
        //        gen_Peridot_retro = gen.put("Peridot Ore retrogen", gen_Peridot_retro)
        //
        gen_MarbleCave = gen.put("Marble Caves", gen_MarbleCave)
        //        gen_MarbleCave_resistance = gen.put("Marble Caves resistance", gen_MarbleCave_resistance)
        //        gen_MarbleCave_retro = gen.put("Marble Caves retrogen", gen_MarbleCave_retro)
        //
        //        gen_Volcano = gen.put("Volcanos", gen_Volcano)
        //        gen_Volcano_resistance = gen.put("Volcano resistance", gen_Volcano_resistance)
        //        gen_Volcano_retro = gen.put("Volcano retrogen", gen_Volcano_retro)
        //
        gen_Copper = gen.put("Copper Ore", gen_Copper)
        //        gen_Copper_resistance = gen.put("Copper Ore resistance", gen_Copper_resistance)
        //        gen_Copper_retro = gen.put("Copper Ore retrogen", gen_Copper_retro)
        //
        gen_Tin = gen.put("Tin Ore", gen_Tin)
        //        gen_Tin_resistance = gen.put("Tin Ore resistance", gen_Tin_resistance)
        //        gen_Tin_retro = gen.put("Tin Ore retrogen", gen_Tin_retro)
        //
        gen_Silver = gen.put("Silver Ore", gen_Silver)
        //        gen_Silver_resistance = gen.put("Silver Ore resistance", gen_Silver_resistance)
        //        gen_Silver_retro = gen.put("Silver Ore retrogen", gen_Silver_retro)
        //
        gen_Electrotine = gen.put("Electrotine Ore", gen_Electrotine)
        //        gen_Electrotine_resistance = gen.put("Electrotine Ore resistance", gen_Electrotine_resistance)
        //        gen_Electrotine_retro = gen.put("Electrotine Ore retrogen", gen_Electrotine_retro)

        //        val compat = new BaseCategory("Compatibility", "Control the loading of various compatibility hooks. These settings are ignored unless the Compatibility module is installed.")
        //        compat_Treecapitator = compat.put("Treecapitator: Gem Axe", compat_Treecapitator, "This allows gem axes to work with treecapitator.")
        //        compat_CCBundledCalbe = compat.put("ComputerCraft: Bundled Cables", compat_CCBundledCalbe, "This allows computers to connect to bundled cables with the RS API.")
        //        compat_Chisel = compat.put("Chisel: Decorative Blocks", compat_Chisel, "Registers ProjectRed decorative blocks with Chisel.")
        //        compat_TConstruct = compat.put("Tinkers Construct: Smeltery", compat_TConstruct, "This adds recipes to the smeltery.")
        //        compat_TExpansion = compat.put("Thermal Expansion: Machine Recipes", compat_TExpansion, "This adds recipes to machines.")
        //        compat_ColoredLights = compat.put("ColoredLights Compat", compat_ColoredLights, "This makes things emit colored light. CLC is in beta state and may cause minor rendering glitches.")
        //        compat_MFRDeepStorage = compat.put("MFR: Deep Storage", compat_MFRDeepStorage, "This allows pipes to recoginze MFR Deep storage units correctly.")
    }
}
